modules/e2ee/OlmAdapter.js

```diff
@@ -470,6 +544,172 @@ export class OlmAdapter extends Listenable {
             }
             break;
         }
+        case OLM_MESSAGE_TYPES.SAS_START: {
+            if (olmData.session) {
+                if (olmData.sas) {
+                    logger.warn(`SAS already created for participant ${pId}`);
+
+                    return;
+                }
+
+                const { uuid } = msg.data;
+
+                // Send ACCEPT.
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_ACCEPT,
+                        data: {
+                            uuid
+                        }
+                    }
+                };
+
+                this._sendMessage(ack, pId);
+            } else {
+                logger.debug(`Received sas init message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-init');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_ACCEPT: {
+            if (olmData.session) {
+                const { commitment, uuid } = msg.data;
+                olmData.sasCommitment = commitment;
+
+                olmData.sas = new Olm.SAS()
+                const pub_key = olmData.sas.get_pubkey();
+
+                // Send KEY.
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_KEY,
+                        data: {
+                            key: pub_key,
+                            isInitializer: true,
```

Comment:
instead of encoding this here each participant should be aware of their own state